### ISSUE TITLE
namespace optimisation ignores variable scopes

### ISSUE THREAD

**Issue by lukeapage** [rollup/rollup#154]
```
// entry
import * as bar from './bar';

function car() {
  var bar = {
    foo: function() {
    }
  };
  bar.foo();
}

// bar.js
export function foo() {};
```

output:
```
function foo() {}

function car() {
  var bar = {
    foo: function() {
    }
  };
  foo();
}
```

expected output:
```
function foo() {}

function car() {
  var bar = {
    foo: function() {
    }
  };
  bar.foo();
}
```

**Comment by lukeapage** [comment]
similarly (and worse/more likely)
```
// entry.js
import * as bar from './bar';

function car() {
  var foo = function() {};
  bar.foo();
}

// bar.js
export function foo() {};
```

results in..
```
function foo() {}

function car() {
  var foo = function() {};
  foo();
}
```

**Comment by lukeapage** [comment]
@Rich-Harris have you considered refactoring rollup into a babel plugin?

advantages:
1. a variable scope management and renaming system for free
2. take advantage of existing plugins, e.g. namespace inlining is a special case of the babel inline plugin, in fact so is inlining the contents of files - you could focus on getting es6 modules into a 100% correct ast and then use inlining plugins to inline and dead code elimination to remove the bits of modules not used
3. most people using es6 need to transpile already. this would avoid reading the ast twice
4. automatic support for every file format babel supports

Sorry if you find the suggestion insulting, rollup is a great package and IMO this technique is the future of dependency management. You are probably well past this stage, but I thought I'd suggest it as even now it could save some work.

**Comment by Rich-Harris** [comment]
@lukeapage not an insult! Though I think you're underestimating the complexity of it :grinning: My sense is that if it was doable, it would have been done, or at least someone would be working on it.

To my knowledge, no-one is. The evidence is that it comes up on Babel's issue tracker repeatedly – https://github.com/babel/babel/issues/495, https://github.com/babel/babel/issues/1118, https://github.com/babel/babel/issues/1681, https://github.com/babel/babel/issues/1806. In fact Sebastian [recommends using Rollup here](https://github.com/babel/babel/issues/1681#issuecomment-136581780).

Having to glue together multiple tools *is* painful. But there are good reasons to want to keep transpilation and bundling separate – for example, if you start to explore ideas like caching (#2), bundle splitting, hot module reloading etc, suddenly you're straying a long way from what Babel is designed for. Also, I've found that while Babel is blazing fast at transpiling individual modules, it gets a bit slower when you're dealing with very large files (such as a large bundle!). I'm not sure where the bottlenecks lie (parsing? analysis? codegen?), but I wouldn't be that surprised if the combination of Babel and Rollup turned out to be as fast or faster than doing the whole job in Babel would be, even without the opportunity for incremental rebuilds. That's because Rollup doesn't have to construct a large AST and generate code from it, it's using a [string replacement parlour trick](https://github.com/rich-harris/magic-string). (A future version of rollup-babel *should* be able to reuse Babel's AST, skipping even that overhead.)

Back to the issue at hand: it does work on the #151 branch, and so probably works on 0.15.0 – is rolling back an option for the time being, or are you dependent on other fixes since then?

**Comment by lukeapage** [comment]
i see,  makes sense.

for this issue i can just wait, I haven't got round to getting rollup
working in production.

**Comment by Victorystick** [comment]
@Rich-Harris I've had this issue in mind while designing the virtual scope class. It should be possible to use it in place of Statement's current implementation, and then do lookups, for namespaces etc, through it.

In the constructor, we'd use the module's local scope as the top scope.
```js
  this.scope = this.module.locals;
```

And in analyse, we'd create new virtual scopes, where we now use the other scope. We'd need to figure out how to manage `var`, since everything else is block scoped in ES2015.
```js
newScope = new Scope( scope );
```

And finally, the namespace optimisation would do its lookups through the scope defined through the AST Node's `_scope` property, as is done now.

```
let scope = this.scope;

// walk( ...
  if ( node._scope ) scope = node._scope.

  // ...

    // At first, we don't have a namespace, so we'll try to look one up.
    const id = scope.lookup( localName );
```

**Comment by Victorystick** [comment]
I'll revisit this after #132.

**Comment by Rich-Harris** [comment]
> We'd need to figure out how to manage `var`, since everything else is block scoped in ES2015.

This is handled in the 'other' `Scope` by passing declarations up to the parent scope until we find a function boundary: https://github.com/rollup/rollup/blob/master/src/ast/Scope.js#L56-L67

Which makes me wonder: why do we have both? It's a little confusing – it seems like their jobs overlap enough that there should just be one `Scope` class. Or would that not make sense?

**Comment by Victorystick** [comment]
Indeed. I hinted at it above.
> ... we'd create new virtual scopes, where we now use the other scope.

After #156 I started looking into replacing `src/ast/Scope` with virtual scopes (`src/Scope`); I'll share what I have once it's closer to working. I only needed to add one or two new methods to `src/Scope`.

Doing so means that we can reuse the same deconfliction logic for the nested scopes. :smiley:

**Comment by Rich-Harris** [comment]
Our shadowing woes appear to be resolved as of 0.18 – closing
